src/ripple/core/Pg.h

```diff
@@ -0,0 +1,503 @@
+#include <ripple/basics/Log.h>
+#include <ripple/protocol/Protocol.h>
+#include <boost/lexical_cast.hpp>
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <libpq-fe.h>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace ripple {
+
+// These postgres structs must be freed only by the postgres API.
+using pg_result_type = std::unique_ptr<PGresult, void (*)(PGresult*)>;
+using pg_connection_type = std::unique_ptr<PGconn, void (*)(PGconn*)>;
+
+/** first: command
+ * second: parameter values
+ *
+ * The 2nd member takes an optional string to
+ * distinguish between NULL parameters and empty strings. An empty
+ * item corresponds to a NULL parameter.
+ *
+ * Postgres reads each parameter as a c-string, regardless of actual type.
+ * Binary types (bytea) need to be converted to hex and prepended with
+ * \x ("\\x").
+ */
+using pg_params =
+    std::pair<char const*, std::vector<std::optional<std::string>>>;
+
+/** Parameter values for pg API. */
+using pg_formatted_params = std::vector<char const*>;
+
+/** Parameters for managing postgres connections. */
+struct PgConfig
+{
+    /** Maximum connections allowed to db. */
+    std::size_t max_connections{std::numeric_limits<std::size_t>::max()};
+    /** Close idle connections past this duration. */
+    std::chrono::seconds timeout{600};
+
+    /** Index of DB connection parameter names. */
+    std::vector<char const*> keywordsIdx;
+    /** DB connection parameter names. */
+    std::vector<std::string> keywords;
+    /** Index of DB connection parameter values. */
+    std::vector<char const*> valuesIdx;
+    /** DB connection parameter values. */
+    std::vector<std::string> values;
+};
+
+//-----------------------------------------------------------------------------
+
+/** Class that operates on postgres query results.
+ *
+ * The functions that return results do not check first whether the
+ * expected results are actually there. Therefore, the caller first needs
+ * to check whether or not a valid response was returned using the operator
+ * bool() overload. If number of tuples or fields are unknown, then check
+ * those. Each result field should be checked for null before attempting
+ * to return results. Finally, the caller must know the type of the field
+ * before calling the corresponding function to return a field. Postgres
+ * internally stores each result field as null-terminated strings.
+ */
+class PgResult
+{
+    // The result object must be freed using the libpq API PQclear() call.
+    pg_result_type result_{nullptr, [](PGresult* result) { PQclear(result); }};
+    std::optional<std::pair<ExecStatusType, std::string>> error_;
+
+public:
+    PgResult()
+    {
+    }
+
+    /** Constructor for successful query results.
+     *
+     * @param result Query result.
+     */
+    PgResult(pg_result_type&& result) : result_(std::move(result))
```

Comment:
Unless there's a motivating reason, this should probably be `explicit`.  I tried it and it compiled as `explicit`.